node-graph/graph-craft/src/proto.rs

```diff
@@ -340,45 +333,59 @@ impl ProtoNetwork {
 	}
 
 	pub fn resolve_inputs(&mut self) {
-		let mut resolved = HashSet::new();
-		while !self.resolve_inputs_impl(&mut resolved) {}
-	}
-	fn resolve_inputs_impl(&mut self, resolved: &mut HashSet<NodeId>) -> bool {
+		// Perform topological sort once
 		self.reorder_ids();
 
-		let mut lookup = self.nodes.iter().map(|(id, _)| (*id, *id)).collect::<HashMap<_, _>>();
-		let compose_node_id = self.nodes.len() as NodeId;
-		let inputs = self.nodes.iter().map(|(_, node)| node.input.clone()).collect::<Vec<_>>();
-		let paths = self.nodes.iter().map(|(_, node)| node.document_node_path.clone()).collect::<Vec<_>>();
-
-		let resolved_lookup = resolved.clone();
-		if let Some((input_node, id, input, mut path)) = self.nodes.iter_mut().filter(|(id, _)| !resolved_lookup.contains(id)).find_map(|(id, node)| {
-			if let ProtoNodeInput::Node(input_node, false) = node.input {
-				resolved.insert(*id);
-				let pre_node_input = inputs.get(input_node as usize).expect("input node should exist");
-				let pre_path = paths.get(input_node as usize).expect("input node should exist");
-				Some((input_node, *id, pre_node_input.clone(), pre_path.clone()))
-			} else {
-				resolved.insert(*id);
-				None
+		let max_id = self.nodes.len() as NodeId - 1;
+
+		// Collect outward edges once
+		let outwards_edges = self.collect_outwards_edges();
```

Comment:
I have to request we use "upstream" or "downstream" since "outwards" conveys no meaning to me and I have no clue if that means upstream or downstream. As a reminder, upstream means easier in the execution, more to the left of the graph. Water flowing through a stream, or data through the graph, progresses downstream and rightward to the output over time.

Reply:
Outwards means the arrow leaves the node and inward means the opposite I've always found that to be quite intuitive, especially considering that we at this point do not yet necessarily have a DAG and without a dag the notion of upstream and downstream doesn't really make sense right?
But this is not a hill I would like to die on

Reply:
To me "outward" implies "radiating outward in all directions" and conveys no meaning to me. Now that you have explained it, I suppose I see your logic for the choice of word but it's not something I would be able to understand without learning and remembering the alternative word choice. Is there a reason that the standard upstream/downstream terminology confuses you?

Reply:
I always mix up upstream and downstream. I think another reason is that there is not direct equivalent term in german so there is not really something to map that against. The notion of not having a direction is actually accurate as we at this stage don't necessarily have an established flow yet, that basically happens after the topological sort. But yeah idk, hypercube chose the terminology initially